solutions/instances/main.tf

```diff
@@ -38,6 +38,26 @@ locals {
   scc_workload_protection_instance_name     = var.prefix != null ? "${var.prefix}-${var.scc_workload_protection_instance_name}" : var.scc_workload_protection_instance_name
   scc_workload_protection_resource_key_name = var.prefix != null ? "${var.prefix}-${var.scc_workload_protection_instance_name}-key" : "${var.scc_workload_protection_instance_name}-key"
   scc_cos_bucket_name                       = var.prefix != null ? "${var.prefix}-${var.scc_cos_bucket_name}" : var.scc_cos_bucket_name
+  create_cross_account_auth_policy          = nonsensitive(!var.skip_cos_kms_auth_policy && var.ibmcloud_kms_api_key != null)
+
+  kms_service_name = var.existing_kms_instance_crn != null ? (
+    can(regex(".*kms.*", var.existing_kms_instance_crn)) ? "kms" : (
+      can(regex(".*hs-crypto.*", var.existing_kms_instance_crn)) ? "hs-crypto" : null
+    )
+  ) : null
+}
+
+# Create IAM Authorization Policy to allow COS to access KMS for the encryption key, if cross account KMS is passed in
+resource "ibm_iam_authorization_policy" "cos_kms_policy" {
+  count                       = local.create_cross_account_auth_policy ? 1 : 0
+  provider                    = ibm.kms
+  source_service_account      = data.ibm_iam_account_settings.iam_account_settings.account_id
+  source_service_name         = "cloud-object-storage"
+  source_resource_instance_id = local.cos_instance_guid
+  target_service_name         = local.kms_service_name
+  target_resource_instance_id = local.existing_kms_guid
+  roles                       = ["Reader"]
+  description                 = "Allow COS instance in the account ${data.ibm_iam_account_settings.iam_account_settings.account_id} to read from the ${local.kms_service_name} instance GUID ${local.existing_kms_guid}"
```

Comment:
When adding an auth policy, you should add a time sleep before trying to deploy the service. For example, see https://github.com/terraform-ibm-modules/terraform-ibm-secrets-manager/blob/dbc64683a586dc0034ce61bc47d66d96cf0a9f3b/main.tf#L74-L79

Once the time sleep exists, you need to add an depends_on it for the call to cos module

Reply:
From what I understand the COS needs to already exist to create this auth policy so adding a sleep would create a cycle. Am I misunderstanding where the depends_on go?

Reply:
@kierramarie The auth policy has to exist before the COS bucket can be created. But the instance has to exist before the auth policy can be created. Unfortunately we are creating both instance and bucket in the COS module call, so I don't think we can apply the time sleep workaround here. If we used a different resource for creating the bucket (such as the buckets submodule), this would not be the case. 

There might be a race condition here though.. what if the bucket creation started to happen before the auth policy was created. We have no way to control that. 
We may need to split bucket creation into its own resource to fix this, but we would need to ensure not to break current consumers.